scipy/stats/_qmc.py

```diff
@@ -0,0 +1,1500 @@
+    collection of subsets of the hypercube, the discrepancy is the greatest
+    absolute difference between the fraction of sample points in one of those
+    subsets and the volume of that subset. There are different definitions of
+    discrepancy corresponding to different collections of subsets. Some
+    versions take a root mean square difference over subsets instead of
+    a maximum.
+
+    A measure of uniformity is reasonable if it satisfies the following
+    criteria [1]_:
+
+    1. It is invariant under permuting factors and/or runs.
+    2. It is invariant under rotation of the coordinates.
+    3. It can measure not only uniformity of the sample over the hypercube,
+       but also the projection uniformity of the sample over non-empty
+       subset of lower dimension hypercubes.
+    4. There is some reasonable geometric meaning.
+    5. It is easy to compute.
+    6. It satisfies the Koksma-Hlawka-like inequality.
+    7. It is consistent with other criteria in experimental design.
+
+    Four methods are available:
+
+    * ``CD``: Centered Discrepancy - subspace involves a corner of the
+      hypercube
+    * ``WD``: Wrap-around Discrepancy - subspace can wrap around bounds
+    * ``MD``: Mixture Discrepancy - mix between CD/WD covering more criteria
+    * ``star``: Star L2-discrepancy - like CD BUT variant to rotation
+
+    Lastly, using ``iterative=True``, it is possible to compute the
+    discrepancy as if we had :math:`n+1` samples. This is useful if we want
+    to add a point to a sampling and check the candidate which would give the
+    lowest discrepancy. Then you could just update the discrepancy with
+    each candidate using `_update_discrepancy`. This method is faster than
+    computing the discrepancy for a large number of candidates.
+
+    References
+    ----------
+    .. [1] Fang et al. Design and modeling for computer experiments,
+       Computer Science and Data Analysis Series, 2006.
+    .. [2] Zhou Y.-D. et al. Mixture discrepancy for quasi-random point sets
+       Journal of Complexity, 29 (3-4) , pp. 283-301, 2013.
+    .. [3] T. T. Warnock. Computational investigations of low discrepancy point
+       sets, Applications of Number Theory to Numerical
+       Analysis, Academic Press, pp. 319-343, 1972.
+
+    Examples
+    --------
+    Calculate the quality of the sample using the discrepancy:
+
+    >>> from scipy.stats import qmc
+    >>> space = np.array([[1, 3], [2, 6], [3, 2], [4, 5], [5, 1], [6, 4]])
+    >>> l_bounds = [0.5, 0.5]
+    >>> u_bounds = [6.5, 6.5]
+    >>> space = qmc.scale(space, l_bounds, u_bounds, reverse=True)
+    >>> space
+    array([[0.08333333, 0.41666667],
+           [0.25      , 0.91666667],
+           [0.41666667, 0.25      ],
+           [0.58333333, 0.75      ],
+           [0.75      , 0.08333333],
+           [0.91666667, 0.58333333]])
+    >>> qmc.discrepancy(space)
+    0.008142039609053464
```

Comment:
I was trying to put together another example that compares the discrepancy of sequences generated by `Sobol`, `Halton`, `np.random.rand`, and `linspace`, but I ran into:
```python3
from scipy.stats.qmc import discrepancy, Sobol, Halton
Sobol(1)
```
Traceback (most recent call last):

  File "<ipython-input-8-9636b540ede7>", line 1, in <module>
    Sobol(d=1)

  File "C:\Users\matth\scipy\scipy\stats\_qmc.py", line 1161, in __init__
    self._scramble()

  File "C:\Users\matth\scipy\scipy\stats\_qmc.py", line 1180, in _scramble
    _cscramble(self.d, ltm, self._sv)

  File "_sobol.pyx", line 286, in scipy.stats._sobol._cscramble

ValueError: Buffer dtype mismatch, expected 'int_t' but got 'long long'
```

Reply:
I was thinking of something like:
```python3
import numpy as np
from scipy.stats.qmc import discrepancy, Sobol, Halton

m = 8  # 2**8 points
d = 2  # 2 dimensions
x1 = Sobol(d, scramble=False).random_base2(m)
x2 = Halton(d, scramble=False).random(2**m)
x3 = np.random.rand(2**m, d)
# the following should be refactored so it's simpler to understand what's going on!
x4 = np.array(np.meshgrid(np.linspace(0, 1, int(2**(m/2))), np.linspace(0, 1, int(2**(m/2))))).T.reshape(-1,2)
print(discrepancy(x1))
print(discrepancy(x2))
print(discrepancy(x3))
print(discrepancy(x4))
```
with interpretation of the results. Do you think that would be interesting / informative? It might be worth mentioning why the grid is the worst of the lot, for example.

Reply:
I don't know about the error. I cannot reproduce on macOS. Also the  @Balandat would you know? 

This could be something to have in a tutorial for instance. Like in the stats tutorial. I would do that in a separate PR if you don't mind. It will take some time to prepare a good tutorial. And I don't want to just copy past my thesis or something else I do in a hurry.

Reply:
@mdhaber My guess is that you are on windows? Apparently cython `int_t` on windows is 32 bit whereas it's 64 bit on unix, which is likely the issue here. Not quite sure why the CI builds worked though... I just pushed a commit that uses `intp_t` instead of `int_t` which should be unambiguous. Can you try if that fixed your issue?

Reply:
Thanks, I am on Windows. I am getting the error in a different place now:
```
(scipydev) C:\Users\matth\scipy>python
Python 3.7.9 (default, Aug 31 2020, 17:10:11) [MSC v.1916 64 bit (AMD64)] :: Anaconda, Inc. on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> from scipy.stats.qmc import discrepancy, Sobol, Halton
>>> Sobol(1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\matth\scipy\scipy\stats\_qmc.py", line 1168, in __init__
    initialize_v(self._sv, d)
  File "_sobol.pyx", line 192, in scipy.stats._sobol.initialize_v
ValueError: Buffer dtype mismatch, expected 'intp_t' but got 'long'
```

Reply:
Hmm ok I think I know what might be going on. Might need to spin up a VM to debug locally.

Reply:
Just pushed another change let's try that again. If it doesn't work I will get serious about the VM.

Reply:
That worked! Thanks. Do you know why that wasn't showing up in CI?

Reply:
It showed up on my previous commit, but not on all the prior runs. Not sure why - maybe some version discrepancies between your numpy/cython and the one in CI?